core/api/serializers/project.py

```diff
@@ -52,8 +66,76 @@ class Meta:
             "plus",
         ]
 
+    def validate_category(self, value):
+        categories = [
+            x[0] for x in ProjectSubmission.ProjectSubmissionCategories.choices
+        ]
+        if value not in categories:
+            raise serializers.ValidationError("Invalid category")
+        return value
+
+
+class ProjectOdsOdpSerializer(serializers.ModelSerializer):
+    """
+    ProjectOdsOdpSerializer class
+    """
+
+    ods_display_name = serializers.SerializerMethodField()
+    ods_substance_id = serializers.IntegerField(required=False)
+    ods_blend_id = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = ProjectOdsOdp
+        fields = [
+            "id",
+            "project_id",
+            "ods_display_name",
+            "odp",
+            "ods_replacement",
+            "co2_mt",
+            "ods_type",
+            "ods_substance_id",
+            "ods_blend_id",
+            "sort_order",
+        ]
+        read_only_fields = ["id"]
+
+    def get_ods_display_name(self, obj):
+        if obj.ods_display_name:
+            return obj.ods_display_name
+        if obj.ods_substance:
+            return obj.ods_substance.name
+        if obj.ods_blend:
+            return obj.ods_blend.name
+        return None
+
+    def validate_ods_substance_id(self, value):
+        substance = Substance.objects.filter(id=value).first()
+        if not substance:
+            raise serializers.ValidationError(f"Invalid ods substance id {value}")
+        return value
+
+    def validate_ods_blend_id(self, value):
+        blend = Blend.objects.filter(id=value).first()
+        if not blend:
+            raise serializers.ValidationError(f"Invalid ods blend id {value}")
+        return value
```

Comment:
Oh, I totally forgot about this yesterday. Indeed using an `IntegerField` field won't perform any validation and would crash the API when wrong values are provided unless we perform validation manually. 

I would suggest switching to [PrimaryKeyRelatedField](https://www.django-rest-framework.org/api-guide/relations/#primarykeyrelatedfield) instead. That should perform all validation required.

---

Alternatively if you prefer using manual validation, I would change the from using `.first()` to `.exists()`. Since that will not load the data from the DB at all, and will issue an `EXISTS` query. 

```python
    def validate_ods_substance_id(self, value):
        if not Substance.objects.filter(id=value).exists():
            raise serializers.ValidationError(f"Invalid ods substance id {value}")
        return value
```